File: Programming Languages/Lab 4/Lexican.java
import java.io.IOException;
import java.nio.file.Files;
import java.nio.file.Path;

public class Lexican {
    static final int EOF = -1;
    static final int LETTER = 0;
    static final int DIGIT = 1;
    static final int UNKNOWN = 99;
    static final int INT_LIT = 10;
    static final int INDENT = 11;
    static final int ASSIGN_OP = 20;
    static final int ADD_OP = 21;
    static final int SUB_OP = 22;
    static final int MULT_OP = 23;
    static final int DIV_OP = 24;
    static final int LEFT_PAREN = 25;
    static final int RIGHT_PAREN = 26;

    static int nextToken;

    static int charClass;
    static char[] lexeme = new char[100];
    static char nextChar;
    static int lexLen;
    static String data = "";

    static int charCounter = 0;

    public static void main(String args[]) {
        Path filename = Path.of("front.txt");

        try {
            data = Files.readString(filename);
        }
        catch (IOException e) {
            System.out.println(e);
            System.exit(1);
        }

        getChar();
        do {
            lex();
        } while (nextToken != EOF);

    }

    public static int lookup(char ch) {
        switch (ch) {
            case '(':
                addChar();
                nextToken = LEFT_PAREN;
                break;
            case ')':
                addChar();
                nextToken = RIGHT_PAREN;
                break;
            case '+':
                addChar();
                nextToken = ADD_OP;
                break;
            case '-':
                addChar();
                nextToken = SUB_OP;
                break;
            case '*':
                addChar();
                nextToken = MULT_OP;
                break;
            case '/':
                addChar();
                nextToken = DIV_OP;
                break;
            default:
                addChar();
                nextToken = EOF;
                break;
        }

        return nextToken;
    }

    public static void addChar() {
        if (lexLen <= 98) {
            lexeme[lexLen++] = nextChar;
            lexeme[lexLen] = 0;
        }
        else {
            System.out.println("Error - lexeme is too long");
        }
    }

    public static void getChar() {
        try {
            nextChar = getc();
            if (Character.isAlphabetic(nextChar)) {
                charClass = LETTER;
            }
            else if (Character.isDigit(nextChar)) {
                charClass = DIGIT;
            }
            else {
                charClass = UNKNOWN;
            }
        }
        catch (IndexOutOfBoundsException e) {
            charClass = EOF;
        }
    }

    public static void getNonBlank() {
        while (nextChar == ' ') {
            getChar();
        }
    }

    public static int lex() {
        lexLen = 0;
        getNonBlank();

        switch (charClass) {
            case LETTER:
                addChar();
                getChar();
                while (charClass == LETTER || charClass == DIGIT) {
                    addChar();
                    getChar();
                }
                nextToken = INDENT;
                break;
            
            case DIGIT:
                addChar();
                getChar();
                while (charClass == DIGIT) {
                    addChar();
                    getChar();
                }
                nextToken = INT_LIT;
                break;
            
            case UNKNOWN:
                lookup(nextChar);
                getChar();
                break;

            case EOF:
                nextToken = EOF;
                lexeme[0] = 'E';
                lexeme[1] = 'O';
                lexeme[2] = 'F';
                lexeme[3] = 0;
                break;
        }

        String lexemeOut = new String(lexeme);

        // Flush lexeme buffer
        for (int i = 0; i < lexeme.length; i++) {
            lexeme[i] = 0;
        }

        System.out.printf("Next token is: %d, Next lexeme is %s\n", nextToken, lexemeOut);
        return nextToken;
    }

    public static char getc(){
        char output = data.charAt(charCounter);
        charCounter++;
        return output;
    }
}

File: Programming Languages/Lab 4/lexican.py
import os

# Definitions
EOF = -1
LETTER = 0
DIGIT = 1
UNKNOWN = 99
INT_LIT = 10
INDENT = 11
ASSIGN_OP = 20
ADD_OP = 21
SUB_OP = 22
MULT_OP = 23
DIV_OP = 24
LEFT_PAREN = 25
RIGHT_PAREN = 26

# Global variables
next_token = 0
char_class = 0
lexeme = ''
next_char = ''
lex_len = 0
data = ''
char_counter = 0

def lookup(ch):
    global next_token
    if ch == '(':
        add_char()
        next_token = LEFT_PAREN
    elif ch == ')':
        add_char()
        next_token = RIGHT_PAREN
    elif ch == '+':
        add_char()
        next_token = ADD_OP
    elif ch == '-':
        add_char()
        next_token = SUB_OP
    elif ch == '*':
        add_char()
        next_token = MULT_OP
    elif ch == '/':
        add_char()
        next_token = DIV_OP
    else:
        add_char()
        next_token = EOF

    return next_token

def add_char():
    global lex_len, lexeme, next_char

    lexeme[lex_len + 1] = next_char
    lexeme[lex_len] = 0

def get_char():
    global next_char, char_class

    next_char = getc()
    if next_char:
        if next_char.isalpha:
            char_class = LETTER
        elif next_char.isdigit:
            char_class
        else:
            char_class = UNKNOWN
    else:
        char_class = EOF

def get_non_blank():
    while not next_char.isspace:
        get_char()

def lex():
    global lex_len, char_class, next_token
    lex_len = 0;
    get_non_blank()

    if char_class == LETTER:
        add_char()
        get_char()
        while char_class == LETTER or char_class == DIGIT:
            add_char()
            get_char()
        next_token = INDENT
    elif char_class == DIGIT:
        add_char()
        get_char()
        while char_class == DIGIT:
            add_char()
            get_char()
        next_token = INT_LIT
    elif char_class == UNKNOWN:
        lookup(next_char)
        get_char()
    elif char_class == EOF:
        next_token = EOF
        lexeme = 'EOF'

    print('Next token is: {}, Next lexeme is: {}'.format(next_token, lexeme))
    return next_token

def getc():
    global char_counter
    output = data[char_counter] if char_counter < len(data) else None
    return output

def main():
    global data, next_token
    with open('front.txt', 'r') as file:
        data = file.read()

        get_char()
        while True:
            lex()

            if next_token == EOF:
                break

if __name__ == '__main__':
    main()